Script to install or update XCP-ng in a VM (#126)
xcp-ng/xcp-ng-tests: README.md

        name: util-linux efitools
```

## install_xcpng.py

This script installs (or upgrades) XCP-ng in a VM using a PXE server whose configuration can be defined dynamically. Basically, it writes files in a directory named after the MAC address of the VM, on a PXE server that will then build a boot configuration for the given MAC address. This is rather specific to Vates' test lab at the moment. If you are interested in automated installation in general, check https://xcp-ng.org/docs/install.html#automated-install.

```
usage: install_xcpng.py [-h] [--installer INSTALLER] host vm_uuid action xcpng_version
```

Installation example:
```
python scripts/test_install_xcpng.py 10.0.0.2 3bdf2cc6-4e6e-526d-4f18-bf7899953af6 install 8.2.1 --installer=https://mirrors.xcp-ng.org/netinstall/8.2.1/
```

Upgrade example:
```
python scripts/test_install_xcpng.py 10.0.0.2 f0f5f010-80c6-25ae-44a2-1fb154e32d14 upgrade 8.2.1
```

The script requires the addressable name or IP of the PXE config server to be defined in `data.py`:
```
# PXE config server for automated XCP-ng installation
PXE_CONFIG_SERVER = 'pxe'
```

> The `installer` parameter is optional. If you leave it empty it will be automatically defined as `http://<PXE_CONFIG_SERVER>/installers/xcp-ng/<version>/`

Reviewer: ```suggestion
The `installer` parameter is optional. If you leave it empty it will be automatically defined as `http://<PXE_CONFIG_SERVER>/installers/xcp-ng/<version>/`.
```

        name: util-linux efitools
```

## install_xcpng.py

> This script install (or upgrade) a VM into a XCP-ng host using a PXE server. You need to have the PXE server configured and running with the proper installer for each version. For more information about PXE boot install you can read the [XCP-ng documentation](https://xcp-ng.org/docs/install.html#pxe-boot-install).

Reviewer: ```suggestion
This script installs (or upgrades) a VM into a XCP-ng host using a PXE server. You need to have the PXE server already configured and running with the proper installer for each version. For more information about PXE boot install you can read the [XCP-ng documentation](https://xcp-ng.org/docs/install.html#pxe-boot-install).
```

> This script install (or upgrade) a VM into a XCP-ng host using a PXE server. You need to have the PXE server configured and running with the proper installer for each version. For more informations about PXE boot install you can read the XCP-ng documentation [here](https://xcp-ng.org/docs/install.html#pxe-boot-install).

Reviewer: ```suggestion
This script install (or upgrade) a VM into a XCP-ng host using a PXE server. You need to have the PXE server configured and running with the proper installer for each version. For more information about PXE boot install you can read the [XCP-ng documentation](https://xcp-ng.org/docs/install.html#pxe-boot-install).
```

The script can display the help with `-h`:

```
usage: install_xcpng.py [-h] [--installer INSTALLER] host vm_uuid action xcpng_version

positional arguments:

>   host                  the IP address or the name of the host where the VM is created.
>   vm_uuid               the UUID of an existing VM to use to create the new host.
>   action                install or upgrade.
>   xcpng_version         the version of xcp-ng.

Reviewer: ```suggestion
  host                  the IP address or the name of the host where the VM is created
  vm_uuid               the UUID of an existing VM to use to create the new host
  action                install or upgrade
  xcpng_version         the version of xcp-ng
```

I know this generated elsewhere, i'll suggest modifications there too.

>   version_xcpng         The version of xcp-ng we wish.

Reviewer: ```suggestion
  version_xcpng         The version of xcp-ng.
```

Reviewer 2: `xcpng_version` instead?

options:
  -h, --help            show this help message and exit
  --installer INSTALLER
                        URL of the installer on the PXE

> ```

Reviewer 2: You use sentences with capital letters and `.` for positional arguments, not here. ;)

Author: i don't understand. If it's about the 'INSTALLER' i don't know where it's coming from, it's not in the code.
i've just paste the return from the console.

Reviewer: you need to change the help print by the script.
- `show this help message and exit` => `Show this help message and exit.`
- `URL of the installer on the PXE` => `URL of the installer on the PXE.`

Author: i don't understand.
i have this :
```
   parser = argparse.ArgumentParser()
    parser.add_argument("host", help="The IP address or the name of the host where the VM is created.")
    parser.add_argument("vm_uuid", help="The UUID of an existing VM to use to create the new host.")
    parser.add_argument("action", metavar='action', choices=['install', 'upgrade'], help="install or upgrade.")
    parser.add_argument("xcpng_version", help="The version of xcp-ng.")
    parser.add_argument("--installer", help="URL of the installer on the PXE")
    args = parser.parse_args()
```
The help is generated automaticly with thehelp=""
i don't know how to change that.

Reviewer: You can edit the installer option description
As for the Help it can be changed with an optional argument IIRC.

Or you can remove all capital letters and `.` for consistency.
We do not have a preference.

Author: but i don't have any capital letter in the code :(

Here an example of the command for an installation:
```
python scripts/test_install_xcpng.py 10.0.0.2 3bdf2cc6-4e6e-526d-4f18-bf7899953af6 install 8.2.1
```

Here another example for an upgrade:
```
python scripts/test_install_xcpng.py 10.0.0.2 f0f5f010-80c6-25ae-44a2-1fb154e32d14 upgrade 8.2.1
```

The script use a variable from the `data.py` file. You need to add the following parameter inside that file.
```
# Default PXE server

> DEF_PXE_IP = '10.0.0.2'

Reviewer: ```suggestion
DEF_PXE_IP = '<IP of your PXE server>'
```

Author: i've use the same model we already have is the model file. So i d'ont think to have a different formalism. I'd prefer to stay coherent with what already exist ;)

Reviewer 3: In the data.py file itself, where the variable is commented, I agree. But in the README at this specific place, Benjamin's suggestion would also be appropriate. Both suit me.

```

> The `installer` parameter isn't mandatory. If you leav it empty it will be created like this: http://<PXE_IP>/installers/xcp-ng/<version>/

Reviewer: ```suggestion
The `installer` parameter isn't mandatory. If you leave it empty it will be created like this: http://<PXE_IP>/installers/xcp-ng/<version>/
```
